ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureCloudDestinationSync.java

```diff
@@ -0,0 +1,374 @@
+      // To avoid overwriting, pass "*" to setIfNoneMatch(String ifNoneMatch)
+      // https://learn.microsoft.com/en-us/java/api/com.azure.storage.blob.blobclient?view=azure-java-stable
+      blobParallelUploadOptions.setRequestConditions(new BlobRequestConditions().setIfNoneMatch("*"));
+      // This is ambry metadata, uninterpreted by azure
+      blobParallelUploadOptions.setMetadata(cloudBlobMetadata.toMap());
+      // Without content-type, get-blob floods log with warnings
+      blobParallelUploadOptions.setHeaders(new BlobHttpHeaders().setContentType("application/octet-stream"));
+
+      BlobContainerClient blobContainerClient = createOrGetBlobStore(blobLayout.containerName);
+      ////////////////////////////////// Upload blob to Azure blob storage ////////////////////////////////////////
+      storageTimer = azureMetrics.blobUploadTime.time();
+      Response<BlockBlobItem> blockBlobItemResponse =
+          blobContainerClient.getBlobClient(blobIdStr)
+              .uploadWithResponse(blobParallelUploadOptions, Duration.ofMillis(cloudConfig.cloudRequestTimeout),
+                  Context.NONE);
+      ////////////////////////////////// Upload blob to Azure blob storage ////////////////////////////////////////
+
+      // Metrics and log
+      // Success rate is effective, Counter is ineffective because it just monotonically increases
+      azureMetrics.blobUploadSuccessRate.mark();
+      // Measure ingestion rate, helps decide fleet size
+      azureMetrics.backupSuccessByteRate.mark(inputLength);
+      logger.debug("Successful upload of blob {} to Azure blob storage with statusCode = {}, etag = {}",
+          blobIdStr, blockBlobItemResponse.getStatusCode(),
+          blockBlobItemResponse.getValue().getETag());
+    } catch (Exception e) {
+      if (e instanceof BlobStorageException
+          && ((BlobStorageException) e).getErrorCode() == BlobErrorCode.BLOB_ALREADY_EXISTS) {
+        // Since VCR replicates from all replicas, a blob can be uploaded by at least two threads concurrently.
+        azureMetrics.blobUploadConflictCount.inc();
+        logger.debug("Failed to upload blob {} to Azure blob storage because it already exists", blobLayout);
+        // We should rarely be here because we get here from replication logic which checks if a blob exists or not before uploading it.
+        // However, we end up here, return true to allow replication to proceed instead of halting it. Else the replication token will not advance.
+        // The blob in the cloud is safe as Azure prevented us from overwriting it due to an ETag check.
+        return true;
+      }
+      azureMetrics.blobUploadErrorCount.inc();
+      String error = String.format("Failed to upload blob %s to Azure blob storage because %s", blobLayout, e.getMessage());
+      logger.error(error);
+      throw new CloudStorageException(error, new StoreException(error, StoreErrorCodes.Unknown_Error));
```

Comment:
Just a note. Since you are wrapping `new StoreException(error, StoreErrorCodes.Unknown_Error)`, you would lose information about `BlobStorageException`. If you plan to use the azure error in caller of this method, please wrap the BlobStorageException itself.

Reply:
thanks, fixed.